src/librustc_mir/interpret/validity.rs

```diff
@@ -580,10 +580,21 @@ impl<'rt, 'mir, 'tcx, M: Machine<'mir, 'tcx>> ValueVisitor<'mir, 'tcx, M>
             }
             ty::Array(tys, ..) | ty::Slice(tys)
                 if {
-                    // This optimization applies only for integer and floating point types
-                    // (i.e., types that can hold arbitrary bytes).
+                    // This optimization applies for types that can hold arbitrary bytes (such as
+                    // integer and floating point types) or for inhabited Zero Sized Types (such as
+                    // the unit type and empty struct types).
                     match tys.kind {
                         ty::Int(..) | ty::Uint(..) | ty::Float(..) => true,
+                        ty::Tuple(t)
+                            if t.len() == 0 || self.ecx.layout_of(tys)?.size == Size::ZERO =>
+                        {
+                            true
+                        }
+                        ty::Adt(adt, _)
+                            if adt.is_struct() && self.ecx.layout_of(tys)?.size == Size::ZERO =>
```

Comment:
Same, this is wrong. Please go back to what you did before.

Also please add a test making sure that we correctly validate things like `(!, !)` and `struct Empty(!);`.